File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/test/java/com/ntnu/gidd/service/token/RefreshTokenServiceImplTest.java
package com.ntnu.gidd.service.token;

import com.ntnu.gidd.config.JwtConfiguration;
import com.ntnu.gidd.exception.RefreshTokenNotFound;
import com.ntnu.gidd.model.RefreshToken;
import com.ntnu.gidd.repository.RefreshTokenRepository;
import com.ntnu.gidd.security.UserDetailsImpl;
import com.ntnu.gidd.security.config.JWTConfig;
import com.ntnu.gidd.security.token.JwtRefreshToken;
import com.ntnu.gidd.security.token.JwtTokenFactory;
import com.ntnu.gidd.security.token.TokenFactory;
import com.ntnu.gidd.util.JwtUtil;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.util.ReflectionTestUtils;

import java.util.Optional;
import java.util.UUID;

import static com.ntnu.gidd.utils.StringRandomizer.getRandomEmail;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.Mockito.*;


@Import({JwtConfiguration.class})
@ExtendWith(MockitoExtension.class)
@ActiveProfiles("test")
class RefreshTokenServiceImplTest {

    @Mock
    private RefreshTokenRepository refreshTokenRepository;

    private RefreshTokenService refreshTokenService;

    private RefreshToken refreshToken;

    private RefreshToken newRefreshToken;

    private JwtRefreshToken jwtRefreshToken;

    private JwtRefreshToken newJwtRefreshToken;

    @BeforeEach
    void setUp() {
        JWTConfig jwtConfig = new JWTConfig();
        TokenFactory tokenFactory = new JwtTokenFactory(jwtConfig);
        JwtUtil jwtUtil = new JwtUtil(jwtConfig);
        refreshTokenService = new RefreshTokenServiceImpl(refreshTokenRepository, jwtUtil);

        ReflectionTestUtils.setField(jwtConfig, "secret", "notasecret");
        ReflectionTestUtils.setField(jwtConfig, "expiration", 24*60*60);
        ReflectionTestUtils.setField(jwtConfig, "refreshExpiration", 24*60*60);

        UserDetailsImpl userDetails = UserDetailsImpl.builder()
                .email(getRandomEmail())
                .build();

        jwtRefreshToken = jwtUtil.parseToken(tokenFactory.createRefreshToken(userDetails).getToken()).get();
        refreshToken = RefreshToken.builder()
                .jti(UUID.fromString(jwtRefreshToken.getJti()))
                .isValid(true)
                .build();

        newJwtRefreshToken = jwtUtil.parseToken(tokenFactory.createRefreshToken(userDetails)
                                                                        .getToken()).get();
        newRefreshToken = RefreshToken.builder()
                .jti(UUID.fromString(newJwtRefreshToken.getJti()))
                .isValid(true)
                .build();

        lenient().when(refreshTokenRepository.save(refreshToken))
                .thenReturn(refreshToken);
        lenient().when(refreshTokenRepository.save(newRefreshToken))
                .thenReturn(newRefreshToken);
        lenient().when(refreshTokenRepository.findById(refreshToken.getJti()))
                .thenReturn(Optional.of(refreshToken));
    }

    /**
     * Test that a refresh token is saved properly.
     */
    @Test
    void testSaveRefreshTokenSavesAndReturnsRefreshToken() {
        RefreshToken actualRefreshToken = refreshTokenService.saveRefreshToken(jwtRefreshToken);

        assertThat(actualRefreshToken).isEqualTo(refreshToken);
    }

    /**
     * Test that the old refresh token is invalidated upon rotation.
     */
    @Test
    void testRotateRefreshTokenInvalidatesTheOldRefreshToken() {
        refreshTokenService.rotateRefreshToken(jwtRefreshToken, newJwtRefreshToken);

        assertThat(refreshToken.isValid()).isFalse();
    }

    /**
     * Test that rotating refresh tokens points the old token to the new.
     */
    @Test
    void testRotateRefreshTokenPointsOldRefreshTokenToNewRefreshToken() {
        refreshTokenService.rotateRefreshToken(jwtRefreshToken, newJwtRefreshToken);

        assertThat(refreshToken.getNext()).isEqualTo(newRefreshToken);
    }

    /**
     * Test that rotating refresh tokens saves the old refresh token.
     */
    @Test
    void testRotateRefreshTokenSavesTheOldRefreshToken() {
        refreshTokenService.rotateRefreshToken(jwtRefreshToken, newJwtRefreshToken);

        verify(refreshTokenRepository, times(1)).save(refreshToken);
    }

    /**
     * Test that the new refresh token is saved as valid upon rotation.
     */
    @Test
    void testRotateRefreshTokenSavesTheNewRefreshTokenAsValid() {
        refreshTokenService.rotateRefreshToken(jwtRefreshToken, newJwtRefreshToken);

        verify(refreshTokenRepository, times(1)).save(newRefreshToken);
    }

    /**
     * Test that the correct refresh token is returned when it exists.
     */
    @Test
    void testGetByJtiWhenRefreshTokenExistsReturnsRefreshToken() {
        RefreshToken actualRefreshToken = refreshTokenService.getByJti(refreshToken.getJti()
                                                                  .toString());

        assertThat(actualRefreshToken).isEqualTo(refreshToken);
    }

    /**
     * Test that an exception is raised when the refresh token does not exist.
     */
    @Test
    void testGetByJtiWhenRefreshTokenDoesNotExistsRaisesException() {
        lenient().when(refreshTokenRepository.findById(refreshToken.getJti()))
                .thenReturn(Optional.empty());

        assertThatExceptionOfType(RefreshTokenNotFound.class).isThrownBy(() -> refreshTokenService.getByJti(refreshToken.getJti()
                                                                                                                    .toString()));
    }

    /**
     * Test that all subsequent tokens are invalidated when invalidating all subsequent tokens.
     */
    @Test
    void testInvalidateSubsequentTokensInvalidatesAllRefreshTokensAfterRefreshToken() {
        refreshToken.setNext(newRefreshToken);

        refreshTokenService.invalidateSubsequentTokens(refreshToken.getJti().toString());

        assertThat(refreshToken.isValid()).isFalse();
        assertThat(newRefreshToken.isValid()).isFalse();
    }
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/test/kotlin/ntnu/idatt2105/security/service/RefreshTokenServiceImplTest.kt
package ntnu.idatt2105.security.service

import io.github.serpro69.kfaker.Faker
import ntnu.idatt2105.security.JwtUtil
import ntnu.idatt2105.security.config.JWTConfig
import ntnu.idatt2105.security.exception.RefreshTokenNotFound
import ntnu.idatt2105.security.repository.RefreshTokenRepository
import ntnu.idatt2105.security.token.JwtRefreshToken
import ntnu.idatt2105.security.token.JwtTokenFactory
import ntnu.idatt2105.security.token.RefreshToken
import ntnu.idatt2105.security.token.TokenFactory
import ntnu.idatt2105.user.service.UserDetailsImpl
import ntnu.idatt2105.user.service.UserDetailsImplBuilder
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.test.context.ContextConfiguration
import java.util.*

@ContextConfiguration(classes = [JWTConfig::class, JwtUtil::class])
@ExtendWith(MockitoExtension::class)
internal class RefreshTokenServiceImplTest {

    @Mock
    private lateinit var refreshTokenRepository: RefreshTokenRepository

    private lateinit var refreshTokenService: RefreshTokenService

    private lateinit var refreshToken: RefreshToken

    private lateinit var newRefreshToken: RefreshToken

    private lateinit var jwtRefreshToken: JwtRefreshToken

    private lateinit var newJwtRefreshToken: JwtRefreshToken

    private val JWT_SECRET_LENGTH: Int = 512
    private val charPool: List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9')

    @BeforeEach
    fun setUp() {

        val randomString = (1..JWT_SECRET_LENGTH)
            .map { kotlin.random.Random.nextInt(0, charPool.size) }
            .map(charPool::get)
            .joinToString("")

        val jwtConfig = JWTConfig(
            uri = "/auth",
            header = "Authorization",
            prefix = "Bearer",
            secret = randomString,
            expiration = 24 * 60 * 60,
            refreshExpiration = 24 * 60 * 60
        )

        val tokenFactory: TokenFactory = JwtTokenFactory(jwtConfig)
        val jwtUtil = JwtUtil(jwtConfig)
        refreshTokenService = RefreshTokenServiceImpl(refreshTokenRepository, jwtUtil)

        val userDetails: UserDetailsImpl = UserDetailsImplBuilder(
            email = Faker().internet.email()
        ).build()

        jwtRefreshToken = jwtUtil.parseToken(tokenFactory.createRefreshToken(userDetails).getToken())!!
        refreshToken = RefreshToken(
            jti = UUID.fromString(jwtRefreshToken.getJti()),
            isValid = true,
            next = null
        )

        newJwtRefreshToken = jwtUtil.parseToken(
            tokenFactory.createRefreshToken(userDetails)
                .getToken()
        )!!

        newRefreshToken = RefreshToken(
            jti = UUID.fromString(newJwtRefreshToken.getJti()),
            isValid = true,
            next = null
        )

        lenient().`when`(refreshTokenRepository.save(refreshToken))
            .thenReturn(refreshToken)
        lenient().`when`(refreshTokenRepository.save(newRefreshToken))
            .thenReturn(newRefreshToken)
        lenient().`when`(refreshTokenRepository.findById(refreshToken.jti))
            .thenReturn(Optional.of(refreshToken))
    }

    @Test
    fun `test that saving refresh token saves and returns token`() {
        val actualRefreshToken = refreshTokenService.saveRefreshToken(jwtRefreshToken)
        assertThat(actualRefreshToken).isEqualTo(refreshToken)
    }

    @Test
    fun `test that rotating refresh token invalidates the old refresh token`() {
        refreshTokenService.rotateRefreshToken(jwtRefreshToken, newJwtRefreshToken)
        assertThat(refreshToken.isValid).isFalse
    }

    @Test
    fun `test that rotating refresh tokens points the old token to the new`() {
        refreshTokenService.rotateRefreshToken(jwtRefreshToken, newJwtRefreshToken)
        assertThat(refreshToken.next).isEqualTo(newRefreshToken)
    }

    @Test
    fun `test that rotating refresh tokens saves the old refresh token`() {
        refreshTokenService.rotateRefreshToken(jwtRefreshToken, newJwtRefreshToken)
        verify(refreshTokenRepository, times(1)).save(refreshToken)
    }

    @Test
    fun `test that the new refresh token is saved as valid upon rotation`() {
        refreshTokenService.rotateRefreshToken(jwtRefreshToken, newJwtRefreshToken)
        verify(refreshTokenRepository, times(1)).save(newRefreshToken)
    }

    @Test
    fun `test that get by jti when the token exists returns the correct refresh token`() {
        val actualRefreshToken = refreshTokenService.getByJti(
            refreshToken.jti
                .toString()
        )
        assertThat(actualRefreshToken).isEqualTo(refreshToken)
    }

    @Test
    fun `test that an exception is raised when the refresh token does not exist`() {
        lenient().`when`(refreshTokenRepository.findById(refreshToken.jti))
            .thenReturn(Optional.empty())

        assertThatExceptionOfType(RefreshTokenNotFound::class.java).isThrownBy {
            refreshTokenService.getByJti(
                refreshToken.jti
                    .toString()
            )
        }
    }

    @Test
    fun `test that all subsequent tokens are invalidated when invalidating all subsequent tokens`() {
        refreshToken.next = newRefreshToken
        refreshTokenService.invalidateSubsequentTokens(refreshToken.jti.toString())

        assertThat(refreshToken.isValid).isFalse
        assertThat(newRefreshToken.isValid).isFalse
    }
}
